src/main/java/teammates/common/datatransfer/questions/FeedbackRubricQuestionDetails.java

```diff
@@ -104,34 +109,49 @@ public boolean extractQuestionDetails(
-                                         boolean hasAssignedWeights) {
-        List<Double> rubricWeights = new ArrayList<>();
+    /**
+     * Extracts Rubric weight values from request parameters map.
+     * @return Returns the list of rubric weights.
+     */
+    private List<List<Double>> getRubricWeightsForEachCell(Map<String, String[]> requestParameters,
+            int numOfRubricChoices, int numOfRubricSubQuestions, boolean hasAssignedWeights) {
+        List<List<Double>> rubricWeightsForEachCell = new ArrayList<>();
 
         if (!hasAssignedWeights) {
-            return rubricWeights;
+            return rubricWeightsForEachCell;
         }
 
-        for (int i = 0; i < numOfRubricChoices; i++) {
-
-            String weight = HttpRequestHelper.getValueFromParamMap(
-                    requestParameters, Const.ParamsNames.FEEDBACK_QUESTION_RUBRIC_WEIGHT + "-" + i);
-            String choice = HttpRequestHelper.getValueFromParamMap(
-                    requestParameters, Const.ParamsNames.FEEDBACK_QUESTION_RUBRIC_CHOICE + "-" + i);
-
-            if (choice == null || weight == null) {
+        int weightRows = -1;
+        for (int i = 0; i < numOfRubricSubQuestions; i++) {
+            String subQn = HttpRequestHelper.getValueFromParamMap(
+                    requestParameters, Const.ParamsNames.FEEDBACK_QUESTION_RUBRIC_SUBQUESTION + "-" + i);
+            if (subQn == null || subQn.trim().isEmpty()) {
+                // No weights should be parsed for a null or empty sub question.
                 continue;
             }
+            boolean rowAdded = false;
 
-            try {
-                rubricWeights.add(Double.parseDouble(weight));
-            } catch (NumberFormatException e) {
-                // Do not add weight to rubricWeights if the weight cannot be parsed
-                log.warning("Failed to parse weight for rubric question: " + weight);
+            for (int j = 0; j < numOfRubricChoices; j++) {
+                String choice = HttpRequestHelper.getValueFromParamMap(
+                        requestParameters, Const.ParamsNames.FEEDBACK_QUESTION_RUBRIC_CHOICE + "-" + j);
+                String paramName = Const.ParamsNames.FEEDBACK_QUESTION_RUBRIC_WEIGHT + "-" + i + "-" + j;
+                String weight = HttpRequestHelper.getValueFromParamMap(requestParameters, paramName);
+                if (weight != null && choice != null) {
+                    if (!rowAdded) {
+                        weightRows++;
+                        rubricWeightsForEachCell.add(new ArrayList<Double>());
+                        rowAdded = true;
+                    }
+                    try {
+                        rubricWeightsForEachCell.get(weightRows).add(Double.parseDouble(weight));
+                    } catch (NumberFormatException e) {
+                        log.warning("Failed to parse weight for rubric question " + weight);
```

Comment:
Do we expect this exception to happen? If no, we should log this at `severe` level so that admin gets an email if it did happen for some unexplained reason.

Reply:
>Do we expect this exception to happen? we should log this at severe level so that admin gets an email if it did happen for some unexplained reason.

Sir, this exception is unlikely to happen, and even if the user by any chance can pass a value that is not parsable (which I don't think will occur as the weight cell only takes numbers as input, and as the weights are required so passing an empty string won't work too), the validateQuestiondetails will catch this and throw an error. So in any case such situations will be handled. 

But if we want to be super cautious here, then I will  modify it to be severe level.

Reply:
Yes, let's use severe: if it shouldn't actually happen, we want to know immediately if it _did_ happen.

Reply:
Okay. On the same note, I will also make the mcq and msq weight log warnings severe.